packages/std/src/math/signed_decimal.rs

```diff
@@ -0,0 +1,3060 @@
+        self.0
+            .checked_rem(other.0)
+            .map(Self)
+            .map_err(|_| DivideByZeroError::new(self))
+    }
+
+    #[must_use = "this returns the result of the operation, without modifying the original"]
+    pub const fn abs_diff(self, other: Self) -> Decimal {
+        Decimal::new(self.0.abs_diff(other.0))
+    }
+
+    #[must_use = "this returns the result of the operation, without modifying the original"]
+    pub fn saturating_add(self, other: Self) -> Self {
+        Self(self.0.saturating_add(other.0))
+    }
+
+    #[must_use = "this returns the result of the operation, without modifying the original"]
+    pub fn saturating_sub(self, other: Self) -> Self {
+        Self(self.0.saturating_sub(other.0))
+    }
+
+    #[must_use = "this returns the result of the operation, without modifying the original"]
+    pub fn saturating_mul(self, other: Self) -> Self {
+        match self.checked_mul(other) {
+            Ok(value) => value,
+            Err(_) => {
+                // both negative or both positive results in positive number, otherwise negative
+                if self.is_negative() == other.is_negative() {
+                    Self::MAX
+                } else {
+                    Self::MIN
+                }
+            }
+        }
+    }
+
+    #[must_use = "this returns the result of the operation, without modifying the original"]
+    pub fn saturating_pow(self, exp: u32) -> Self {
+        match self.checked_pow(exp) {
+            Ok(value) => value,
+            Err(_) => {
+                // odd exponent of negative number results in negative number
+                // everything else results in positive number
+                if self.is_negative() && exp % 2 == 1 {
+                    Self::MIN
+                } else {
+                    Self::MAX
+                }
+            }
+        }
+    }
+
+    /// Converts this decimal to a signed integer by rounding down
+    /// to the next integer, e.g. 22.5 becomes 22 and -1.2 becomes -2.
+    ///
+    /// ## Examples
+    ///
+    /// ```
+    /// use core::str::FromStr;
+    /// use cosmwasm_std::{SignedDecimal, Int128};
+    ///
+    /// let d = SignedDecimal::from_str("12.345").unwrap();
+    /// assert_eq!(d.to_int_floor(), Int128::new(12));
+    ///
+    /// let d = SignedDecimal::from_str("-12.999").unwrap();
+    /// assert_eq!(d.to_int_floor(), Int128::new(-13));
+    ///
+    /// let d = SignedDecimal::from_str("-0.05").unwrap();
+    /// assert_eq!(d.to_int_floor(), Int128::new(-1));
+    /// ```
+    #[must_use]
+    pub fn to_int_floor(self) -> Int128 {
+        if self.is_negative() {
+            // Using `x.to_int_floor() = -(-x).to_int_ceil()` for a negative `x`,
+            // but avoiding overflow by implementing the formula from `to_int_ceil` directly.
+            let x = self.0;
+            let y = Self::DECIMAL_FRACTIONAL;
+            if x.is_zero() {
+                Int128::zero()
```

Comment:
This line will never get executed. The guarding condition to get to this line is: `number is negative and equal to zero`. I did't found a representation for negative zeros in `Int128`, e.g. `SignedDecimal::from_str("-0.0")` always returns positive zero.

Reply:
Great catch! Integers and signed decimals should not have negative zeros. This is a float thing only.

Reply:
Great catch indeed!